### PR TITLE
[WIP]Seata Agent

### PR DESCRIPTION
**Pull request by kezhenxu94**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues
#2523

### CODE REVIEW

**Comment by kezhenxu94** [comment]
Drafting this to get comments on the idea of supporting Seata, there are many chores to do (such as license headers)

**Comment by wu-sheng** [comment]
Let's work on design of how seata plugin working on both tc and tm. Then license and module structure, which are very easy to have results.

**Comment by kezhenxu94** [comment]
I've tested some scenarios and here are what the traces look like (will append to this comment when more scenarios tests are conducted), it may not cover all the possible scenarios, if I've missed anything, please point it out, thanks :)

- normal scenario (commit)

normal

- roll back

roll back

- TC server is down

tc down

**Comment by wu-sheng** [comment]
@kezhenxu94 The rollback trace is not very clear to me. Could you explain more?  In your scenario, which service triggers the rollback? I expect you could show the trace, one service fails, then TM/TC works together to make other services rollback too.

**Comment by kezhenxu94** [comment]
> @kezhenxu94 The rollback trace is not very clear to me. Could you explain more? In your scenario, which service triggers the rollback? I expect you could show the trace, one service fails, then TM/TC works together to make other services rollback too.

yes I'll elaborate more on this, in this example:

```java
    @Override
    @GlobalTransactional(timeoutMills = 300000, name = "dubbo-demo-tx")
    public void purchase(String userId, String commodityCode, int orderCount) {
        LOGGER.info("purchase begin ... xid: " + RootContext.getXID());
        storageService.deduct(commodityCode, orderCount);
        if (orderCount % 2 == 0) {
            throw new IllegalArgumentException("xxx");
        }
    }
```

- the `business.purchase` throws exception after calling `storageService.deduct` but before finished the global transaction

- the exception triggers the `TransactionManager` (TM) to request (from TC) a rollback

- the `TransactionCoordinator` (TC) requests all the `ResourceManager`s to roll back

below the highlighted span is where the `RM` does the rollback operation (by doing some compensative operations) as the `TC` demanded

image
image

**Comment by kezhenxu94** [comment]
will begin to write test cases and interceptors' unit tests

**Comment by wu-sheng** [comment]
How does topology work? Does it show relationship of TMs and TC clear?
Optional plugin requires extra document.

**Comment by wu-sheng** [comment]
Also, prepare document to ask seata team to review, from header manipulate, trace and topology points. They are really authority reviewers. SkyWalking team including integration tests are just designed for checking codes run as expected.

I will treat their approve as the most important key to merge this

**Comment by wu-sheng** [comment]
Recommend to do documents and seata team review first.

File: apm-sniffer/apm-sdk-plugin/seata-plugin/pom.xml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    <packaging>jar</packaging>
+
+    <artifactId>seata-plugin</artifactId>
+
+    <properties>
+        <seata.version>0.6.0</seata.version>
+    </properties>
+
+    <dependencies>
+        <dependency>
```
Comment (slievrly) [inline apm-sniffer/apm-sdk-plugin/seata-plugin/pom.xml]: Seat-all as an all in one jar, can be used on all client side

File: apm-sniffer/apm-sdk-plugin/seata-plugin/pom.xml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        <groupId>org.apache.skywalking</groupId>
+        <version>6.2.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+    <packaging>jar</packaging>
+
+    <artifactId>seata-plugin</artifactId>
+
+    <properties>
+        <seata.version>0.6.0</seata.version>
```
Comment (slievrly) [inline apm-sniffer/apm-sdk-plugin/seata-plugin/pom.xml]: It is recommended to upgrade to 0.6.1

**Comment by CoffeeLatte007** [comment]
Some point were left out:
-handleRetryRollbacking
-handleRetryCommitting
-handleAsyncCommitting
-LockManager

**Comment by kezhenxu94** [comment]
> Some point were left out:
> -handleRetryRollbacking
> -handleRetryCommitting
> -handleAsyncCommitting
> -LockManager

Hi @CoffeeLatte007 , thanks for reviewing, the `LockManager` instrumentation is added(as the trace below.

As for the three `handleXxx` methods, they are indeed not instrumented, but the underlying operations (`BranchCommit`, `BranchRollback`) are instrumented, therefore the interactions between TC and TM/RM are in trace links.

I think it's enough to show the interactions between TC and TM/RM, instrumenting the three methods just adds some local spans, without more detailed information?

Please correct me if I'm wrong. Thanks :-)


SkyWalking

File: apm-sniffer/apm-sdk-plugin/seata-plugin/pom.xml (hunk 2 above)
Comment (kezhenxu94) [reply apm-sniffer/apm-sdk-plugin/seata-plugin/pom.xml]: Thanks, done

File: apm-sniffer/apm-sdk-plugin/seata-plugin/pom.xml (hunk 1 above)
Comment (kezhenxu94) [reply apm-sniffer/apm-sdk-plugin/seata-plugin/pom.xml]: Thanks, done

**Comment by CoffeeLatte007** [comment]
> > Some point were left out:
> > -handleRetryRollbacking
> > -handleRetryCommitting
> > -handleAsyncCommitting
> > -LockManager
> 
> Hi @CoffeeLatte007 , thanks for reviewing, the `LockManager` instrumentation is added(as the trace below.
> 
> As for the three `handleXxx` methods, they are indeed not instrumented, but the underlying operations (`BranchCommit`, `BranchRollback`) are instrumented, therefore the interactions between TC and TM/RM are in trace links.
> 
> I think it's enough to show the interactions between TC and TM/RM, instrumenting the three methods just adds some local spans, without more detailed information?
> 
> Please correct me if I'm wrong. Thanks :-)
> 
> SkyWalking

The underlying operations are used in many methods. We don't know which method is using it.

**Comment by slievrly** [comment]
image

**Comment by wu-sheng** [comment]
Interesting, how this NPE happens.

**Comment by zhaoyuguang** [comment]
Look like `RmRpcClient.getInstance()`  has no create skyWalkingDynamicField;
so peerAddress is null, ConcurrentHashMap::get(null)  NPE?

**Comment by asfgit** [comment]
FAILURE

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/skywalking-e2e/15/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/skywalking-e2e/30/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/skywalking-ci2/3/

**Comment by zhaoyuguang** [comment]
@slievrly @kezhenxu94 @wu-sheng  I will pick this relate issue up

**Pull request closed without merging**